index.js

```diff
@@ -1,13 +1,197 @@
+import {spawn} from 'node:child_process';
+import {once, on} from 'node:events';
+import {stripVTControlCharacters} from 'node:util';
+import path from 'node:path';
 import process from 'node:process';
-import {normalizeArguments, getCommand, spawnProcess} from './spawn.js';
-import {addPromiseMethods} from './iterable.js';
-
-export default function nanoSpawn(file, commandArguments, options) {
-	[commandArguments, options] = normalizeArguments(commandArguments, options);
-	const context = {
-		start: process.hrtime.bigint(),
-		command: getCommand(file, commandArguments),
-	};
-	const resultPromise = spawnProcess(file, commandArguments, options, context);
-	return addPromiseMethods(resultPromise, context);
+import {fileURLToPath} from 'node:url';
+import {lineIterator, combineAsyncIterators} from './iterable.js';
+import {getForcedShell, escapeArguments} from './windows.js';
+import {handlePipe} from './pipe.js';
+
+export default function nanoSpawn(first, second = [], third = {}) {
+	let [file, previous] = Array.isArray(first) ? first : [first, {}];
+	let [commandArguments, options] = Array.isArray(second) ? [second, third] : [[], second];
+
+	const start = previous.start ?? process.hrtime.bigint();
+	const spawnOptions = getOptions(options);
+	const command = [previous.command, getCommand(file, commandArguments)].filter(Boolean).join(' | ');
+	const context = {start, command};
+	[file, commandArguments] = handleNode(file, commandArguments);
+	const input = getInput(spawnOptions);
+
+	const instancePromise = getInstance(file, commandArguments, spawnOptions, context);
+	const resultPromise = Object.assign(getResult(instancePromise, input, context), {nodeChildProcess: instancePromise});
+	const finalPromise = previous.resultPromise === undefined ? resultPromise : handlePipe(previous, resultPromise);
+
+	const stdoutLines = lineIterator(finalPromise, 'stdout');
+	const stderrLines = lineIterator(finalPromise, 'stderr');
+	return Object.assign(finalPromise, {
+		stdout: stdoutLines,
+		stderr: stderrLines,
+		[Symbol.asyncIterator]: () => combineAsyncIterators(stdoutLines, stderrLines),
+		pipe: (file, second, third) => nanoSpawn([file, {...context, resultPromise: finalPromise}], second, third),
```

Comment:
When `.pipe()` calls `nanoSpawn()` recursively, it passes a `previous` object with information from the previous subprocess (its promise, its command string and its start time). That object is empty is `.pipe()` is not used.